src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -96,9 +96,23 @@ protected void ShutdownConnectedNodes(List<NodeContext> contextsToShutDown, bool
             // Send the build completion message to the nodes, causing them to shutdown or reset.
             _processesToIgnore.Clear();
 
+            // We wait for child nodes to exit to avoid them changing the terminal
+            // after this process terminates.
+            bool waitForExit =  !enableReuse &&
+                                !Console.IsInputRedirected &&
+                                Traits.Instance.EscapeHatches.EnsureStdOutForChildNodesIsPrimaryStdout;
+
             foreach (NodeContext nodeContext in contextsToShutDown)
             {
-                nodeContext?.SendData(new NodeBuildComplete(enableReuse));
+                if (nodeContext is null)
+                {
+                    continue;
+                }
+                nodeContext.SendData(new NodeBuildComplete(enableReuse));
+                if (waitForExit)
+                {
+                    nodeContext.WaitForExit();
```

Comment:
This seems like it's waiting for each child process individually before telling the next one to shut down. Could we send all the shut down messages, then wait for them all instead?

Reply:
Good point, this should be concurrent.
In addition, the RPS tests for VS shutdown times might get triggered by this, even when concurrent. Nodes should be waited only when MSBuild is run from the cmdline. Couple of ways to detect this:
- `AssemblyUtilities.EntryAssembly` is MSBuild.{exe, dll}. Hopefully it doesn't appear as dotnet on .net core :). If you go down this path might as well add a new `bool BuildEnvironmentHelper.IsRunningInCommandLine` (also tests)
- Don't wait when `BuildEnvironment.RunningInVisualStudio == true`. But this will ignore any other process that's using the BuildManager APIs to build.

Sadly I can't think of a straightforward way to test this.

Reply:
> Nodes should be waited only when MSBuild is run from the cmdline. 

I think the `!Console.IsInputRedirected` may be covering this requirement?

Reply:
> Could we send all the shut down messages, then wait for them all instead?

I've changed to concurrent waits.